ui/selectors/selectors.js

```diff
@@ -477,43 +477,96 @@ export function getSelectedAccountNativeTokenCachedBalanceByChainId(state) {
+ * including the native tokens, without hardcoding any native token information.
  *
  * @param {object} state - Redux state
- * @returns {object} An array of tokens with balances for the given account. Data relationship will be chainId => balance
+ * @returns {object} An object mapping chain IDs to arrays of tokens (including native tokens) with balances.
  */
 export function getSelectedAccountTokensAcrossChains(state) {
   const { allTokens } = state.metamask;
-  const { address: selectedAddress } = getSelectedInternalAccount(state);
+  const selectedAddress = getSelectedInternalAccount(state).address;
 
-  // Initialize an empty object to hold tokens by chainId for the selected account
   const tokensByChain = {};
 
-  // Loop through each chain (0x1, 0x89, etc.)
-  for (const chainId in allTokens) {
-    if (allTokens[chainId]) {
-      const chainData = allTokens[chainId];
+  const nativeTokenBalancesByChainId =
+    getSelectedAccountNativeTokenCachedBalanceByChainId(state);
 
-      // Check if the selected account exists within this chain
-      if (chainData[selectedAddress]) {
-        // Ensure the chainId key exists in tokensByChain
-        if (!tokensByChain[chainId]) {
-          tokensByChain[chainId] = [];
-        }
+  const chainIds = new Set([
+    ...Object.keys(allTokens || {}),
+    ...Object.keys(nativeTokenBalancesByChainId || {}),
+  ]);
 
-        // Add each token to the array under its chainId
-        chainData[selectedAddress].forEach((token) => {
-          const tokenWithChain = { ...token, chainId };
-          tokensByChain[chainId].push(tokenWithChain);
-        });
-      }
+  chainIds.forEach((chainId) => {
+    if (!tokensByChain[chainId]) {
+      tokensByChain[chainId] = [];
     }
-  }
+
+    if (allTokens[chainId] && allTokens[chainId][selectedAddress]) {
+      allTokens[chainId][selectedAddress].forEach((token) => {
+        const tokenWithChain = { ...token, chainId, isNative: false };
+        tokensByChain[chainId].push(tokenWithChain);
+      });
+    }
+
+    const nativeBalance = nativeTokenBalancesByChainId[chainId];
+    if (nativeBalance) {
+      const nativeTokenInfo = getNativeTokenInfo(state, chainId);
+      tokensByChain[chainId].push({
+        ...nativeTokenInfo,
+        // TODO: infer this value
+        address: '',
+        balance: nativeBalance,
+        chainId,
+        isNative: true,
+      });
+    }
+  });
 
   return tokensByChain;
 }
 
+/**
+ * Retrieves native token information (symbol, decimals, name) for a given chainId from the state,
+ * without hardcoding any values.
+ *
+ * @param {object} state - Redux state
+ * @param {string} chainId - Chain ID
+ * @returns {object} Native token information
+ */
+function getNativeTokenInfo(state, chainId) {
+  const { networkConfigurationsByChainId } = state.metamask;
+
+  const networkConfig = networkConfigurationsByChainId?.[chainId];
+
+  if (networkConfig) {
+    const symbol = networkConfig.nativeCurrency || 'NATIVE';
```

Comment:
Instead of `NATIVE`, use `AssetType.native`

Reply:
https://github.com/MetaMask/metamask-extension/pull/28193/commits/9b07669164b36339042117f935a70e8c2b57c16f